### ISSUE TITLE
Make it possible to configure images

### ISSUE THREAD

**Issue by aarnq** [elastisys/compliantkubernetes-apps#1884]
**Is your feature request related to a problem? Please describe.**

Compliant Kubernetes is static when it comes to the configuration of which images it uses, and we should change this to address some key scenarios:

- Using a private registry instead of a public registry
- Using a pull through cache
- Patching an image version to mitigate vulnerabilites

**Describe the solution you'd like**

Make it configurable for each instance of where we are using an image so we can:
- Configure the registry of the image (falling back to the one defined by the chart)
- Configure the repository of the image (falling back to the one defined by the chart)
- Configure the tag of the image (falling back to the one defined by the chart)

Make it configurable on a global level so we can:
- Configure the registry of all images (if they do not have an specific override).

This should also include images in resources created by operators to ensure that they are captured as well.

**Definition of done:**

- [ ] Images used can be individually configured.
- [ ] Image registry can be globally configured.

**Comment by cristiklein** [comment]
We solved this by configuring the container runtime, as opposed to the `apps` layer: https://elastisys.io/compliantkubernetes/operator-manual/air-gapped-environment/#migrating-images-manually

Is this issue still relevant?

**Comment by simonklb** [comment]
> We solved this by configuring the container runtime, as opposed to the `apps` layer: https://elastisys.io/compliantkubernetes/operator-manual/air-gapped-environment/#migrating-images-manually
> 
> Is this issue still relevant?

@aarnq feel free to correct me if I'm wrong. I think the idea here is for us to declaratively configure each image so that we are in more control of what is running in the cluster. It looks to me that the link you provided describes how to ingest images when you don't want to pull directly from the upstream registries.

**Comment by aarnq** [comment]
> > We solved this by configuring the container runtime, as opposed to the `apps` layer: https://elastisys.io/compliantkubernetes/operator-manual/air-gapped-environment/#migrating-images-manually
> > Is this issue still relevant?
> 
> @aarnq feel free to correct me if I'm wrong. I think the idea here is for us to declaratively configure each image so that we are in more control of what is running in the cluster. It looks to me that the link you provided describes how to ingest images when you don't want to pull directly from the upstream registries.

Yes, so the goal of this issue is to go further and allow us to have an inventory of images that we can approve by pinning the checksum and sign. Allowing us to secure our supply chain by fetching the images we need, and better respond to vulnerabilities by using rebuilt or patched images dynamically, rather than statically.
And it will help if we want to start to rebuild images to ensure that they have the latest patches applied when we prepare releases.

And it would compliment the air-gapped scenario as it would provide a better inventory of images that are required, and potentially simplifies how that can be managed in apps.
(In such scenarios configuring mirrors in the container runtime will still be needed as this doesn't cover the Kubernetes and infrastructure layer, but it will provide a process to implement it elsewhere.)

**Comment by cristiklein** [comment]
@aarnq Understood, thanks for the clarification!

**Comment by aarnq** [comment]
> Just a quick clarification: by "repository" we mean the part that comes _after_ the registry, but _before_ the last "/" in the image specification, so for example:
> 
> ```
> my.registry.domain.com/foo/bar/image-name:v1.2.3@sha256:babafacecaca
> ```
> 
> would parse into:
> 
>     * registry: `my.registry.domain.com`
> 
>     * repository: `foo/bar`
> 
>     * image: `image-name`
> 
>     * tag: `v1.2.3`
> 
>     * digest: `sha256:babafacecaca`
> 
> 
> correct?

Correct.

**Comment by rarescosma** [comment]
Another quick question: what should happen if an image specifier includes its own repository, but we also set the global one to enabled? i.e.

```
container_uri: repo-here/image-name-here:v1.2.3
global:
  repository:
    enabled: true
    uri: the-global-repo
```

Should we ignore the global repo completely and just go with `repo-here/image-name-here:v1.2.3` or should we _prepend_ the global one to the image specified one, i.e. `the-global-repo/repo-here/image-name-here:v1.2.3` ?

Thinking about this in the context of pull-through caches where we'd probably want to have a single global registry, but then use repo fragments to specify the source registries.

**Comment by aarnq** [comment]
> Another quick question: what should happen if an image specifier includes its own repository, but we also set the global one to enabled? i.e.
> 
> ```
> container_uri: repo-here/image-name-here:v1.2.3
> global:
>   repository:
>     enabled: true
>     uri: the-global-repo
> ```
> 
> Should we ignore the global repo completely and just go with `repo-here/image-name-here:v1.2.3` or should we _prepend_ the global one to the image specified one, i.e. `the-global-repo/repo-here/image-name-here:v1.2.3` ?
> 
> Thinking about this in the context of pull-through caches where we'd probably want to have a single global registry, but then use repo fragments to specify the source registries.

The core of this feature is to provide an explicit image list, and ensure that what's configured there is what is actually used in the manifests.

In the context of pull-through caches, this feature isn't supposed to be how we configure proxy caches, as that is something to be done via the container runtime.
Instead this is supposed to serve as the information about _which_ proxy caches we need to setup.

In the context of explicit image mirroring / vendoring this is supposed to override the registry and repository, and point to a location we setup with all the images in one place.

This use of overriding registry and repository is used for explicit mirroring or vendoring of images, as we then point to a location where we have explicitly made them available.
Alternatively the override repository could be prepended if we want to maintain the upstream repository name.

I will review you PR either later today or tomorrow.
